### PR TITLE
Restore interpreter debug output

### PR DESCRIPTION
**Pull request by johanandren**
Refs #22433

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 1]
```diff
@@ -296,8 +299,7 @@ final class GraphInterpreter(
         logic.preStart()
       } catch {
         case NonFatal(e) ⇒
-          e.printStackTrace()
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: good, I noticed on more `printStackTrace` in `tryAbort` the other day, should that really be a `printStackTrace`

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala (hunk 1 above)
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Must be a leftover, I'll remove that

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 2]
```diff
@@ -351,10 +353,7 @@ final class GraphInterpreter(
         def reportStageError(e: Throwable): Unit = {
           if (activeStage == null) throw e
           else {
-            // TODO: Get error reporting back
-            //val stage = stages(activeStage.stageId)
-
-            log.error(e, "Error in stage [{}]: {}", activeStage, e.getMessage)
+            log.error(e, "Error in stage [{}]: {}", activeStage.originalStage.getOrElse(activeStage), e.getMessage)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Hm this won't show exception type though which is likely important, add `{}: {}` (e, ex.getMessage)?

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: We pass the exception as first parameter as well, wont that lead to the stack trace being written and all? Actually maybe the getMessage is duplication even, @drewhk had removed and I reintroduced that after looking at the old logic.

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: the first `e` param should be enough for the exception type, but I think adding the message in the message is good

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Ah, I completely missed the first parameter. All's good then here.

**Review by ktoso — APPROVED** [review]
LGTM, merge at will

**Comment by drewhk** [comment]
wait, let me review...

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: leftovers, sorry

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: Looks good

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 3]
```diff
@@ -631,45 +630,48 @@ final class GraphInterpreter(
...
-  //        case _ ⇒
-  //      }
-  //      builder.append("\n")
-  //    }
-  //
-  //    builder.append("}\n")
-  //    builder.append(s"// $queueStatus (running=$runningStages, shutdown=${shutdownCounter.mkString(",")})")
-  //    builder.toString()
-  //  }
+  override def toString: String = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: thanky thanks!

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 4]
```diff
@@ -233,6 +234,13 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
 
   /**
    * INTERNAL API
+   *
+   * If possible a link back to the stage that the logic was created with, used for debugging.
+   */
+  private[stream] var originalStage: OptionVal[GraphStageWithMaterializedValue[_ <: Shape, _]] = OptionVal.None
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: yes, I wanted to propose this. While this inflates the memory usage of the stage, we had array fields before that stored the stages anyway.

**Review by drewhk — APPROVED** [review]
LGTM

I think there are some debug information lost in the ActorGraphInterpreter, too. Look up the diff compared to master.

**Comment by johanandren** [comment]
Ah, `GraphInterpreterShell.toString` is still missing. I'm on it.

**Comment by drewhk** [comment]
And I think the toStrings for the Upstream and Downstream boundaries can get some love, too. One of them can print the actual port of the stage next to it, but the other does not right now (don't remember which).

**Comment by johanandren** [comment]
Sample output for a `GraphInterpreterShell`:

```
 GraphInterpreterShell(
    logics: [
      Merge attrs: [InputBuffer(4,16), Dispatcher(akka.test.stream-dispatcher), SupervisionStrategy(<function1>), Name(merge)],
      ActorOutputBoundary(port=Merge.out(656757897), demand=0, finished=false) attrs: [],
      BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false) attrs: [],
      BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false) attrs: []
    ],
    connections: [
      Connection(0, 0, akka.stream.scaladsl.Merge$$anon$2@1d13beec, 2, BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false), akka.stream.scaladsl.Merge$$anon$2$$anonfun$9$$anon$9@58daa8c8, BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false), 8, Empty),
      Connection(1, 0, akka.stream.scaladsl.Merge$$anon$2@1d13beec, 3, BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false), akka.stream.scaladsl.Merge$$anon$2$$anonfun$9$$anon$9@358fe39a, BatchingActorInputBoundary(forPort=akka.stream.scaladsl.Merge$$anon$2@1d13beec, fill=0/16, completed=false, canceled=false), 8, Empty),
      Connection(2, 1, ActorOutputBoundary(port=Merge.out(656757897), demand=0, finished=false), 0, akka.stream.scaladsl.Merge$$anon$2@1d13beec, ActorOutputBoundary(port=Merge.out(656757897), demand=0, finished=false), akka.stream.scaladsl.Merge$$anon$2@1d13beec, 1, Empty)
    ]
  )
```

**Comment by johanandren** [comment]
rebased and fingers crossed

**Comment by johanandren** [comment]
GraphDSLDocSpec and GraphDSLDocTest failing, unrelated to this PR.

**Pull request merged**